### ISSUE TITLE
random spaces added with jinja macro

### ISSUE THREAD

**Issue by rjake** [sqlfluff/sqlfluff#1248]
sqlfluff is awesome! I noticed the following issues when trying to use `sqlfluff fix` I'm new to sqlfluff, jinja and dbt. I would've simplified the column names and such but it seemed the longer they were, the more the issue revealed itself. The `with` statement also seems necessary for the issue to occur.

## Expected Behaviour
the usual beautiful code

## Observed Behaviour
random spaces

![image](https://user-images.githubusercontent.com/14003618/128551084-66091724-782a-4e2a-b20e-303a8268141c.png)

## Steps to Reproduce
It seems to be related to the `{%- ... %}` syntax
 ```sql
{%- set value_lookup = [
   ('123456', 'score_health'),
   ('234567', 'score_pain'),
   ('345678', 'score_home')
] %}

with
   questions as (
       select
           visit_key,
           form_question_id,
       {%- for value_id, value_name in value_lookup %}
           max(
               case
                   when
                       form_question_id = '{{ value_id }}'
                       then answer_as_numeric
               end
           ) as {{ value_name }},
       {%- endfor %}
       min(answer_date) as answer_date
       from
           {{ ref('question_patient_answered') }}
       group by
           visit_key,
           form_question_id
   )
select
   questions.visit_key,
   questions.form_question_id,
{%- for value_id, value_name in value_lookup %}
   questions.{{ value_name }},
{%- endfor %}
   questions.answer_date
from
   questions

```

```
== [test.sql] FAIL
L:  11 | P:  29 | L008 | Commas should be followed by a single whitespace unless followed by a comment.
L:  19 | P:  25 | L008 | Commas should be followed by a single whitespace unless followed by a comment.
L:  26 | P:  16 | L018 | WITH clause closing bracket should be aligned with WITH keyword.
L:  30 | P:  13 | L008 | Commas should be followed by a single whitespace unless followed by a comment.
L:  32 | P:  12 | L008 | Commas should be followed by a single whitespace unless | followed by a comment.
==== fixing violations ====
5 fixable linting violations found
```

## Version
sqlfluff:  0.6.2 
Python: 3.7.9

## Configuration
```
[sqlfluff]
# Need to ignore templating until we can use the dbt templater
ignore = templating
exclude_rules = L003,L004,L046,L013,L022,L027,L031,L033,L034,L035,L036,L037,L039

[sqlfluff:templater:jinja:context]
target = 'dev'

[sqlfluff:indentation]
indented_joins = True

[sqlfluff:rules]
max_line_length = 115

[sqlfluff:rules:L010]
capitalisation_policy = lower
```

**Comment by barrywhart** [comment]
I have confirmed this still occurs in the latest `main`, i.e. this commit: `e74bf2069406922905347fcaca007167b1f4f63d`. I'll try and triage this (not necessarily trying to fix it at this time).

As an experiment, I looked at the lint output (before applying fixes) and saw 4 violations of L008 and 1 violation of L018. L018 relates to the `WITH` statement. I also noticed that the extra spaces and newlines were all added near these reported issues. Since the issue description mentions that it only occurs if a `WITH` statement is present, I tested with a modified `.sqlfluff` that _only_ runs L018:
```
rules = L018
#exclude_rules = L003,L004,L046,L013,L022,L027,L031,L033,L034,L035,L036,L037,L039
```

With this modified configuration, I still see the issue around line 26 but not the others:
```
      group by
          visit_key,
          form_
question_id
  )
```

This may be a better way to _begin_ investigating the issue, since it only involves a single lint rule.

**Comment by barrywhart** [comment]
I'm able to reproduce the same behavior with a simplified query:
```
{%- set value_lookup = [
  ('123456', 'score_health')
] %}

with
  questions as (
      select
          visit_key,
      {%- for value_id, value_name in value_lookup %}
          max(a),
      {%- endfor %}
      min(answer_date) as answer_date
      from
          {{ ref('question_patient_answered') }}
      group by
          form_question_id
  )
select * from t
```

I noticed a few things:
* `.sqlfluff` is set to `ignore = templating`
* In the Python code, `L018` specifies `_works_on_unparsable = False`
* `sqlfluff parse` does not report a parse error for this query. Should it?

In the `sqlfluff parse` output for this simplified SQL, the `ref()` syntax disappears and the actual `table_reference` shows as `question_patient_answered`. Is this the intended behavior?

```
[L: 12, P: 31]      |                        [META] dedent:
[L: 12, P: 31]      |                        from_clause:
[L: 12, P: 31]      |                            keyword:                          'from'
[L: 12, P: 35]      |                            newline:                          '\n'
[L: 12, P: 36]      |                            whitespace:                       '          '
[L: 13, P:  8]      |                            from_expression:
[L: 13, P:  8]      |                                [META] indent:
[L: 13, P:  8]      |                                from_expression_element:
[L: 13, P:  8]      |                                    table_expression:
[L: 13, P:  8]      |                                        table_reference:
[L: 13, P:  8]      |                                            identifier:       'question_patient_answered'
[L: 14, P: 35]      |                                [META] dedent:
[L: 14, P: 35]      |                        newline:                              '\n'
[L: 14, P: 36]      |                        whitespace:                           '      '
[L: 14, P: 42]      |                        groupby_clause:
[L: 14, P: 42]      |                            keyword:                          'group'
[L: 14, P: 47]      |                            whitespace:                       ' '
[L: 14, P: 48]      |                            keyword:                          'by'
[L: 15, P:  1]      |                            [META] indent:
[L: 15, P:  1]      |                            newline:                          '\n'
[L: 15, P:  2]      |                            whitespace:                       '          '
[L: 15, P: 12]      |                            column_reference:
[L: 15, P: 12]      |                                identifier:                   'form_question_id'
[L: 16, P: 13]      |                            [META] dedent:
[L: 16, P: 13]      |                    newline:                                  '\n'
[L: 16, P: 14]      |                    whitespace:                               '  '
[L: 16, P: 16]      |                    [META] dedent:
[L: 16, P: 16]      |                    end_bracket:                              ')'
[L: 16, P: 17]      |            newline:                                          '\n'
```

**Comment by barrywhart** [comment]
@alanmcruickshank: Any thoughts on this issue? Are there other open issues like this (incorrect position info in the presence of Jinja templating)?

If you're busy, do you have any thoughts on what area of the code to begin the investigation?

**Comment by pwildenhain** [comment]
Thanks for hopping on this @barrywhart you rock!

**Comment by tunetheweb** [comment]
@barrywhart Your PR #1364 seems to fix this one too. The For loop indentation is still a little funny but at least it doesn't corrupt the SQL once that fix is merged.

**Comment by barrywhart** [comment]
@rjake: This should be fixed in version 0.6.5.
